src/contracts/libraries/GPv2Encoding.sol

```diff
@@ -355,23 +356,30 @@ library GPv2Encoding {
 
     /// @dev Decodes an interaction from calldata into memory.
     ///
-    /// An encoded interaction has three components: the target address, the data
-    /// length, and the actual interaction data of variable size.
+    /// An encoded interaction has four components: the target address, the
+    /// optional call value, the data length, and the actual interaction data of
+    /// variable size.
     ///
     /// ```
     /// struct EncodedInteraction {
     ///     address target;
+    ///     bytes1 hasValue;
     ///     uint24 dataLength;
+    ///     uint256? value;
     ///     bytes callData;
     /// }
     /// ```
     ///
     /// All entries are tightly packed together in this order in the encoded
-    /// calldata. Example:
+    /// calldata. Examples:
     ///
-    /// input:    0x73c14081446bd1e4eb165250e826e80c5a523783000010000102030405060708090a0b0c0d0e0f
-    /// decoding:   [...............target.................][leng][............data..............]
-    /// stride:                                          20     3    (defined in length field) 16
+    /// input:    0x73c14081446bd1e4eb165250e826e80c5a523783_00_000010_____000102030405060708090a0b0c0d0e0f
+    /// decoding:   [...............target.................][hv][leng][val][............data..............]
+    /// stride:                                          20   1     3    0    (defined in length field) 16
+    ///
+    /// input:    0x0000000000000000000000000000000000000000_01_0000000000000000000000000000000000000000000000000000000de0b6b3a7640000
+    /// decoding:   [...............target.................][hv][leng][............................value.............................][data]
+    /// stride:                                          20   1     3                                                              32     0
```

Comment:
Instead of using a full byte for a boolean flag we could use it to encode the length of `value`. The case `length = 0` is equivalent to `hv = 0`, the case `length = 32` to `hv = 1`, `length > 32` is invalid, and in the case `0 < length < 32` we save some gas in the encoding. The tradeoff is that the code is arguably more complicated.

Reply:
Interesting idea... I don't think it is much more complicated actually.

Reply:
The one thing I dislike is the ambiguity in the encoding. This is already the case though since `hv=0` is the same as `hv=1 and value = 0` and all non-zero `hv` are treated the same.

Reply:
Encodings are not signed and there is no issue if two interactions are repeating, so I'm quite ok with multiple encodings for the same interaction. Currently this is also the case for our orders, since we have two signature schemes!

Reply:
Any reason we don't include an example for the case `hv = 0`? Or do we plan not to use it?

Reply:
@bh2smith the first example uses `hasValue = 0`.

Reply:
Except that `hv = 1`

![Screenshot 2021-01-13 at 10 19 18](https://user-images.githubusercontent.com/11778116/104431985-d311bb80-5588-11eb-8b71-80dccf863d60.png)

Reply:
That's the byte size of `hasValue` (which is always 1 byte). The value is `0x00`